lib/fluent/plugin/in_tail/position_file.rb

```diff
@@ -48,56 +55,89 @@ def unwatch(path)
         end
       end
 
-      def self.parse(file)
-        compact(file)
+      def load
+        compact
 
-        file_mutex = Mutex.new
         map = {}
-        file.pos = 0
-        file.each_line {|line|
-          m = POSITION_FILE_ENTRY_REGEX.match(line)
-          unless m
-            $log.warn "Unparsable line in pos_file: #{line}"
-            next
+        @file_mutex.synchronize do
+          @file.pos = 0
+
+          @file.each_line do |line|
+            m = POSITION_FILE_ENTRY_REGEX.match(line)
+            next if m.nil?
+
+            path = m[1]
+            pos = m[2].to_i(16)
+            ino = m[3].to_i(16)
+            seek = @file.pos - line.bytesize + path.bytesize + 1
+            map[path] = FilePositionEntry.new(@file, @file_mutex, seek, pos, ino)
           end
-          path = m[1]
-          pos = m[2].to_i(16)
-          ino = m[3].to_i(16)
-          seek = file.pos - line.bytesize + path.bytesize + 1
-          map[path] = FilePositionEntry.new(file, file_mutex, seek, pos, ino)
-        }
-        new(file, file_mutex, map)
+        end
+
+        @map = map
+      end
+
+      # This method is similer to #compact but it tries to get less lock to avoid a lock contention
+      def try_compact
+        last_modified = nil
+        size = nil
+
+        @file_mutex.synchronize do
+          last_modified = @file.mtime
+          size = @file.size
+        end
+
+        entries = fetch_compacted_entries
```

Comment:
having this function outside of the lock is probably a race as it reads the file without a lock and as far as I can tell the file writes are not atomic
so this can read a file currently being modified by another thread leading to corrupted data

probable cause of: https://github.com/fluent/fluentd/issues/2918

Reply:
This code seems not to have nothing to do with #2918 because [here](https://github.com/fluent/fluentd/pull/2805/files/8946e536e48c8bda125f7fd075f7a0c388b8850a#diff-0f4f3ff9be203ad2654dd6d9ee3a1f7bR93) verify that the file content doesn't change.
So there is no need to lock here.